### PR TITLE
Automated GitHub releases

### PR DESCRIPTION
**Pull request by bastimeyer**
See https://github.com/streamlink/streamlink/issues/261#issuecomment-268732799 and https://github.com/streamlink/streamlink/pull/310#issuecomment-267738808

This adds support for automated Github releases once a new new tag gets pushed to the repository.
The .tar.gz source archive and the PGP signature file are currently not being uploaded and need to be added in another commit...

When a new tag gets pushed, TravisCI automatically creates a new Github release (the `RELEASES_API_KEY` env var needs to be set first in the TravisCI settings, didn't want to include this in the `.travis.yml` file), uploads the Windows installer and then executes a python script that updates the release `name` and `body` based on the tag and format of the `CHANGELOG.md` file. This means that the format can't be changed, or the script needs to be altered. The script is based on [this](https://github.com/streamlink/streamlink-twitch-gui/blob/v1.0.1/build/travis/update_release.js) (which has been working great so far for me). I'm not a python dev and this was my first attempt, so things will very likely not be perfect. 😱

Tag validation can also be added in the future, so that it only creates releases when a signed tag was successfully validated.

See a successful upload on my test account/repo:
https://github.com/bastimeyer-test/my-test-repo2/compare/0.2.0...0.2.1
https://travis-ci.org/bastimeyer-test/my-test-repo2/jobs/186876690
https://github.com/bastimeyer-test/my-test-repo2/releases/tag/0.2.1

### CODE REVIEW

**Comment by gravyboat** [comment]
Good stuff @bastimeyer! I'm going to leave this open in case others want to review but it gets a 👍 from me.

**Comment by beardypig** [comment]
The installer build script needs to be updated so that it builds with the simple version number (without the commit hash) on tags. At the moment it will copy the installer with `x.y.z+abcdef` as the version. 

Creating the release page is fine, as you said, it still needs to have some extra files uploaded for it to be complete though.

Maybe the release can be marked as pre-release until all the build artifacts are added and we can do some manual verification? 

At some point we wanted to change the release notes to rst, so they are compatible with the docs (also pypi), maybe it's a good time to do that? 

The script is is fine, not super pythonic, but OK ;-)

**Comment by bastimeyer** [comment]
> so that it builds with the simple version number

It already does...
https://github.com/streamlink/streamlink/blob/0.2.0/script/makeinstaller.sh#L16-L17

> pre-release

Sure. `prerelease` simply needs to be set to `true` in the http PATCH request.

> rst

So the difference here are the headers, right? I'm not very familiar with that format. But yeah, that's a simple change...

**Comment by beardypig** [comment]
Ah,  true it does build :-D

Yeah, just the headers for rst vs. md in our case, I think.

**Comment by bastimeyer** [comment]
Updated the PR...
It is now based on the rst changelog PR #271 (which hasn't been merged yet, but has been approved)

Here are the updated URLs to prove that it's working correctly (fake 0.2.2 release):
https://github.com/bastimeyer-test/my-test-repo2/compare/e92bc9812efc40b39e69f2f554bd007975b68c60...0.2.2
https://travis-ci.org/bastimeyer-test/my-test-repo2/jobs/191032083
https://github.com/bastimeyer-test/my-test-repo2/releases/tag/0.2.2

**Comment by beardypig** [comment]
Nice! Now if we just automate the tar + asc (could use a travis encrypted pgp key) we are sorted :-)

**Comment by bastimeyer** [comment]
Updated it again, so that the header is not included anymore in the release page text...

**Comment by bastimeyer** [comment]
Someone needs to set the `RELEASES_API_KEY` env var in Travis first...

**Comment by cdrage** [comment]
Thanks for all the hard work @bastimeyer ! :D Makes releases *sooooo* much easier.

So I'm assuming that all it takes is a new tag on GitHub + notes added to it? How do we trigger this in the TravisCI? Assuming Travis automatically checks periodically and initiates a build?

File: .travis.yml [hunk 1]
```diff
@@ -36,15 +36,29 @@ addons:
...
-  access_key_id:
-    secure: sjjc1tM756TvFbD8+0VCs/MOLFRhY+VZv/lq7pvNldtbGdv21lcJguvgg0gUwUZ7SihceyL1YagfUlDU2vWS1N5Si04z0VUcHO4lbj9nUJ4CXy27pMNFuk92HmblxYQ4bJw/xVqzsrT9Zeh4tIrbV+F3lYXtg7FDqjpm5bqrUJ5ULyWbhax31s2k8JRVnG2fxVmJiU+j4zwgzCqeZJXQkvXsfdzSe88mLyIrl7Otfjlp1rySocxStQ8qlmkvRAmsYmsKriUq+VgtX21HFAIJqHQp4WbL15eh0aFFzbmORf7OMIKmsmkyeYYeM6xp+Jm9m7ZcWHZ+7siplVM/9ksmumN3+CGp6pogduZ5zFH4BtRQa9o8N2SWnPiPvMrpoyLbHbQhg/uhcMJkYI1yrzj/s9MEH7DVnDEMmgc/UKsyTC2PRwZcgqYYPHxqRVWXzvBWDUe1wtxdNyaXWWKdD/1E7u8XF2+zPTPaUNAYAF1rSi5GNF38uRNMC+QSuEm8wZ/32KXFN4m+RW3NtpT9YHI8MCo4ofVxpRUMcEUwPDF/jV5zW7Krz7LEeL5/zcHMcNay/Ls8e3eApQShAcLAx1jZo+EjznVGXtlRbH9ABXmPO5AcxUH5+20tMiMoj0q6hJen03X8Jx+OMmcDNDufDgPikSym3KVC3Lvj65lG4G9pAyY=
+  - provider: releases
+    api_key: "${RELEASES_API_KEY}"
+    file: "${STREAMLINK_INSTALLER_DIST_DIR}/streamlink-${TRAVIS_TAG}.exe"
+    file_glob: true
+    skip_cleanup: true
+    on:
+      tags: true
+      condition: $BUILD_INSTALLER = yes
```
Comment (cdrage) [inline .travis.yml]: So on release (new tags) this would build Windows. Mind adding a comment to the travis CI file? :)

File: .travis.yml [hunk 2]
```diff
@@ -36,15 +36,29 @@ addons:
...
+  - provider: script
+    script: python script/github_releases.py
+    skip_cleanup: true
+    on:
+      tags: true
+      condition: $BUILD_INSTALLER = yes
```
Comment (cdrage) [inline .travis.yml]: Here would be the Linux builds right (another comment in the travis-ci would help)

File: .travis.yml [hunk 3]
```diff
@@ -36,15 +36,29 @@ addons:
...
+    access_key_id:
+      secure: sjjc1tM756TvFbD8+0VCs/MOLFRhY+VZv/lq7pvNldtbGdv21lcJguvgg0gUwUZ7SihceyL1YagfUlDU2vWS1N5Si04z0VUcHO4lbj9nUJ4CXy27pMNFuk92HmblxYQ4bJw/xVqzsrT9Zeh4tIrbV+F3lYXtg7FDqjpm5bqrUJ5ULyWbhax31s2k8JRVnG2fxVmJiU+j4zwgzCqeZJXQkvXsfdzSe88mLyIrl7Otfjlp1rySocxStQ8qlmkvRAmsYmsKriUq+VgtX21HFAIJqHQp4WbL15eh0aFFzbmORf7OMIKmsmkyeYYeM6xp+Jm9m7ZcWHZ+7siplVM/9ksmumN3+CGp6pogduZ5zFH4BtRQa9o8N2SWnPiPvMrpoyLbHbQhg/uhcMJkYI1yrzj/s9MEH7DVnDEMmgc/UKsyTC2PRwZcgqYYPHxqRVWXzvBWDUe1wtxdNyaXWWKdD/1E7u8XF2+zPTPaUNAYAF1rSi5GNF38uRNMC+QSuEm8wZ/32KXFN4m+RW3NtpT9YHI8MCo4ofVxpRUMcEUwPDF/jV5zW7Krz7LEeL5/zcHMcNay/Ls8e3eApQShAcLAx1jZo+EjznVGXtlRbH9ABXmPO5AcxUH5+20tMiMoj0q6hJen03X8Jx+OMmcDNDufDgPikSym3KVC3Lvj65lG4G9pAyY=
```
Comment (cdrage) [inline .travis.yml]: Annddd nightlies! Awesome!

**Comment by cdrage** [comment]
@bastimeyer Ahhh. I see. So the release.sh would *still* be required. But the Windows .exe is automated for generation.

File: .travis.yml (hunk 2 above)
Comment (cdrage) [inline .travis.yml]: Ah. So this just updates the changelog via the script. Nothing to do with source code upload for Linux.

**Comment by bastimeyer** [comment]
The Windows installer [gets built on every commit](https://github.com/bastimeyer/streamlink/blob/e386ea3276dbca357d01946be9d5594ce7911bc7/.travis.yml#L28) (because of the nightlies).

The TravisCI deploy config just checks for every tag being push and then creates/updates the github release for this tag ([and uploads the release files](https://github.com/bastimeyer/streamlink/blob/e386ea3276dbca357d01946be9d5594ce7911bc7/.travis.yml#L41)).

Without a changelog, the release will fail, so it needs to be added before pushing the tag. The tag name has to be equal to the version from the changelog header.

**Comment by bastimeyer** [comment]
See here (a fake release with my alt account):
https://github.com/bastimeyer-test/my-test-repo2/commit/f655b9ba0012273c14c3049adc0ae25dfd23a7ba

**Comment by beardypig** [comment]
I added a release api key to travis which should have sufficient access....

**Comment by gravyboat** [comment]
What else are we waiting on to merge this and https://github.com/streamlink/streamlink/pull/369 ?

**Comment by beardypig** [comment]
@gravyboat I think it's good to go. You can merged just this one and close #369 if you want, this one has the commits from #369 :-)

**Comment by gravyboat** [comment]
Good stuff, thanks @bastimeyer, @beardypig, and @cdrage!

**Comment by stevek123** [comment]
Now that it's easier to create releases, are you going to create a new one any time soon?

**Pull request merged**